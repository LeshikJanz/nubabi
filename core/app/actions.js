// @flow
import axios from 'axios';
import { Observable } from 'rxjs/Observable';
import "../shared/observable";

import type {
  Action,
  AppOnlineAction,
  AppErrorAction,
  OnAuthAction,
  Deps,
  FirebaseUser,
  ResetSettingsAction,
} from '../types';

export const appStarted = (): AppStartedAction => ({
  type: 'APP_STARTED',
});

export const appError = (error: Object): AppErrorAction => ({
  type: 'APP_ERROR',
  payload: error,
  error: true,
});

export const appOnline = (online: boolean): AppOnlineAction => ({
  type: 'APP_ONLINE',
  payload: { online },
});

export const appSuccess = (message: string): AppSuccessAction => ({
  type: 'APP_SUCCESS',
  payload: message,
});

export const appInfo = (message: string): AppInfoAction => ({
  type: 'APP_INFO',
  payload: message,
});

export const onAuth = (user: ?FirebaseUser, token?: string): OnAuthAction => ({
  type: 'ON_AUTH',
  payload: { user, token },
});

export const resetSettings = (): ResetSettingsAction => ({
  type: 'RESET_SETTINGS',
});

const appStartedEpic = (action$: any, deps: Deps) => {
  const { firebaseAuth } = deps;

  const onAuth$ = Observable.create(observer => {
    const unsubscribe = firebaseAuth().onAuthStateChanged(firebaseUser => {
      if (firebaseUser) {
        firebaseAuth()
          .currentUser.getToken()
          .then(token => {
            if (__DEV__) {
              axios
                .post('http://localhost:8080/graphql-config', {
                  token: `Bearer ${token}`,
                })
                .catch(() => {});
            }
            observer.next(onAuth(firebaseUser, token));
          });
      } else {
        observer.next(onAuth(firebaseUser));
      }
    });
    return unsubscribe;
  });

  const streams: Array<any> = [onAuth$];

  return action$
    .filter((action: Action) => action.type === 'APP_STARTED')
    .mergeMap(() => Observable.merge(...streams));
};

export const resetSettingsEpic = (action$: any) => {
  return action$
    .filter(action => action.type === 'RESET_SETTINGS')
    .mapTo(appSuccess('Settings has been reset'));
};

export const epics = [appStartedEpic, resetSettingsEpic];
