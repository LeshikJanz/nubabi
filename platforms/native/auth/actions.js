// @flow
import { LoginManager, AccessToken } from 'react-native-fbsdk';

export const loginWithFacebook = () => {
  return new Promise((resolve, reject) => {
    LoginManager.logInWithReadPermissions(['public_profile', 'email']).then(
      result => {
        if (result.isCancelled) {
          reject(new Error('Facebook login was cancelled'));
        } else {
          AccessToken.getCurrentAccessToken()
            .then(data => {
              resolve(data);
            })
            .catch(error => reject(error));
        }
      },
      error => {
        reject(error);
      },
    );
  });
};
