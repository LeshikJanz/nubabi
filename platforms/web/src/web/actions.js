// @flow
import { Action } from 'core/types/types';
import { Baby } from 'core/types/modelTypes';

export function getBabySuccess(baby: Baby): Action {
  return {
    type: 'GET_BABY_SUCCESS',
    payload: baby,
  };
}

export function getBabyFailure(err): Action {
  return {
    type: 'GET_BABY_FAILURE',
    payload: err,
    error: true,
  };
}
