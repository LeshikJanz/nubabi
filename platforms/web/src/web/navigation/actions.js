// @flow
import type { Deps } from "../../common/types";
import { Observable } from "rxjs";
import history from "web/navigation/history";

export const resetNavigation = (routeName: string, index?: number = 0) => {
  return history.push(routeName);
};

export const navigate = (routeName: string, params: Object = {}) => {
  return history.push(routeName);
};

const resetNavigationEpic = (action$: any, deps: Deps) => {
  console.log(deps);
  console.log("RESET");
  return action$
    .ofType("NAVIGATION_RESET")
    .mergeMap(action =>
      Observable.of(
        resetNavigation(action.payload.routeName, action.payload.index)
      )
    );
};

export const epics = [resetNavigationEpic];
