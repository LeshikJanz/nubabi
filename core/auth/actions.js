// @flow
import type { Action, AuthProvider, AuthProviderData, Deps } from '../types';
import { Observable } from 'rxjs/Observable';
import { prop } from 'ramda';
import { resetNavigation } from '../navigation/actions';

export function loginRequest(
  emailOrProviderData: string | AuthProviderData,
  passwordOrProvider: string | AuthProvider,
): Action {
  let providerData;
  let provider: AuthProvider;

  if (typeof emailOrProviderData === 'string') {
    if (__DEV__) {
      // eslint-disable-next-line no-console
      console.warn(
        "DEPRECATED: loginRequest should take arguments in form loginRequest(providerData, provider). Invocations with (email, password) will be removed in the next major version. Use loginRequest({ email, password }, 'email') instead.",
      );
    }

    providerData = {
      email: emailOrProviderData,
      password: passwordOrProvider,
    };
    provider = 'email';
  } else {
    providerData = emailOrProviderData;
    // $FlowFixMe$
    provider = passwordOrProvider;
  }

  // $FlowFixMe$
  return {
    type: 'LOGIN_REQUEST',
    payload: providerData,
    meta: {
      provider,
    },
  };
}

export function loginSuccess({
  email,
  uid,
}: {
  email: string,
  uid: string,
}): Action {
  return {
    type: 'LOGIN_SUCCESS',
    payload: { email },
    meta: {
      token: uid,
    },
  };
}

export function loginFailure(err: Error): Action {
  return {
    type: 'LOGIN_FAILURE',
    payload: err,
    error: true,
  };
}

export function logout(): Action {
  return {
    type: 'LOGOUT',
  };
}

export const signInWithEmailAndPassword = (firebaseAuth, action) => {
  const { email, password } = action.payload;
  return firebaseAuth().signInWithEmailAndPassword(email, password);
};

export const getProviderClass = (provider: string) => {
  const providerName = provider.toUpperCase();
  switch (providerName) {
    case 'FACEBOOK':
      return 'FacebookAuthProvider';
    default:
      return null;
  }
};

const signInWithProvider = (firebaseAuth, action) => {
  return new Promise((resolve, reject) => {
    const { provider: providerName } = action.meta;
    const provider = getProviderClass(providerName);
    let accessToken;

    if (providerName === 'facebook') {
      accessToken = prop('accessToken', action.payload);
    }

    if (!provider) {
      reject(new Error('Tried to authenticate with unknown provider'));
    }

    if (!accessToken) {
      reject(
        new Error('An access token was not retrieved. Sign in has failed.'),
      );
    }

    const credential = firebaseAuth[provider].credential(accessToken);
    firebaseAuth()
      .signInWithCredential(credential)
      .then(user => {
        // Only allow signups on web by specifying `allowSignup` meta property
        // to loginRequest.
        // We should have 2 providers here.
        if (action.meta.allowSignup !== true && user.providerData.length < 2) {
          user.delete();
          reject(
            new Error(
              "We couldn't find an existing user account with the information provided",
            ),
          );
        }

        resolve(user);
      })
      .catch(reject);
  });
};

const loginEpic = (action$: any, { firebaseAuth }: Deps) => {
  return action$
    .filter((action: Action) => action.type === 'LOGIN_REQUEST')
    .mergeMap(action => {
      const promise =
        action.meta.provider === 'email'
          ? signInWithEmailAndPassword
          : signInWithProvider;

      // $FlowFixMe$
      const signIn = Observable.from(promise(firebaseAuth, action)).map(
        firebaseUser => loginSuccess(firebaseUser),
      );

      return Observable.merge(
        signIn,
        action$
          .ofType('APP_ONLINE')
          .withLatestFrom(action$.ofType('GET_BABIES_SUCCESS'))
          .take(1)
          .mapTo(resetNavigation('home')),
      ).catch(err => Observable.of(loginFailure(err)));
    });
};

const logoutEpic = (action$: any, { firebaseAuth }: Deps) => {
  return action$
    .filter((action: Action) => action.type === 'LOGOUT')
    .mergeMap(() => Observable.fromPromise(firebaseAuth().signOut()))
    .mapTo(resetNavigation('login'));
};

export const epics = [loginEpic, logoutEpic];
