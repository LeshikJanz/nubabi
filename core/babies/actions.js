// @flow
import type {
  Action,
  SelectBabyAction,
  DeleteBabyAction,
  MutationResultAction,
} from '../types';
import { Observable } from 'rxjs/Observable';
import { path } from 'ramda';
import { gql } from 'react-apollo';
import { appError, appSuccess } from '../app/actions';
import { resetNavigation } from '../navigation/actions';
import formatPossessive from '../helpers/formatPossessive';
import { removeEdge } from '../../libs/graphql-utils';

export function setBabyPhoto(photoAsDataUrl: string) {
  return {
    type: 'SET_BABY_PHOTO',
    payload: photoAsDataUrl,
  };
}

export function selectBaby(id: string): SelectBabyAction {
  return {
    type: 'SELECT_BABY',
    payload: id,
  };
}

export function deleteBaby(id: string): DeleteBabyAction {
  return {
    type: 'DELETE_BABY',
    payload: id,
  };
}

export function getBabiesRequest(): Action {
  return {
    type: 'GET_BABIES_REQUEST',
  };
}

export function getBabiesSuccess(babies): Action {
  return {
    type: 'GET_BABIES_SUCCESS',
    payload: babies,
  };
}

export function getBabiesFailure(err): Action {
  return {
    type: 'GET_BABIES_FAILURE',
    payload: err,
    error: true,
  };
}

const createBabyEpic = (action$: any) => {
  return action$
    .filter((action: MutationResultAction) => {
      return (
        action.type === 'APOLLO_MUTATION_RESULT' &&
        action.operationName === 'CreateBaby'
      );
    })
    .mergeMap(({ result: { data } }) => {
      if (data.createBaby) {
        return [
          selectBaby(data.createBaby.edge.node.id),
          resetNavigation('home'),
        ];
      }

      return Observable.of(null);
    });
};

const deleteBabyEpic = (action$: any, { apollo }: Deps) => {
  return action$
    .ofType('DELETE_BABY')
    .mergeMap(action => {
      return Observable.fromPromise(
        apollo.mutate({
          mutation: gql`
            mutation DeleteBaby($input: DeleteBabyInput!) {
              deleteBaby(input: $input) {
                edge {
                  node {
                    id
                  }
                }
              }
            }
          `,
          variables: {
            input: {
              id: action.payload,
            },
          },
          update: (store, data) => {
            const id = path(['data', 'deleteBaby', 'edge', 'node', 'id'], data);
            if (!id) {
              return;
            }
            const query = gql`
              query {
                viewer {
                  babies {
                    edges {
                      node {
                        id
                        name
                        avatar {
                          url
                        }
                      }
                    }
                  }
                }
              }
            `;

            const prev = store.readQuery({ query });
            const newData = removeEdge(
              ['viewer', 'babies', 'edges'],
              id,
              prev,
              data,
            );

            if (newData) {
              store.writeQuery({ query, data: newData });
            }
          },
        }),
      );
    })
    .mergeMap(() => {
      return Observable.fromPromise(
        apollo.query({
          query: gql`
            query ChooseFirstBaby {
              viewer {
                babies(first: 1) {
                  edges {
                    node {
                      id
                      name
                    }
                  }
                }
              }
            }
          `,
        }),
      );
    })
    .mergeMap(data => {
      const newBaby = path(
        ['data', 'viewer', 'babies', 'edges', '0', 'node'],
        data,
      );

      const newBabyId = newBaby ? newBaby.id : null;
      const messageSuffix = newBaby
        ? `You're now viewing ${formatPossessive(newBaby.name)} profile.`
        : "You'll need to select or create a baby next.";

      const message = 'Baby was successfully deleted.';
      return [
        selectBaby(newBabyId),
        resetNavigation('home'),
        appSuccess(`${message} ${messageSuffix}`),
      ];
    })
    .catch(() =>
      Observable.of(
        appError(new Error('There was a problem deleting this baby.')),
      ),
    );
};

export const epics = [createBabyEpic, deleteBabyEpic];
