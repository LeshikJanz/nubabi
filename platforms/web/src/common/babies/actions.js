import { Observable } from "rxjs/Observable";
import { Action, MutationResultAction } from "../types";
import { resetNavigation } from "../navigation/actions";

export function selectBaby(id): Action {
  return {
    type: "SELECT_BABY",
    payload: id
  };
}

export function getBabiesRequest(): Action {
  return {
    type: "GET_BABIES_REQUEST"
  };
}

export function getBabiesSuccess(babies): Action {
  return {
    type: "GET_BABIES_SUCCESS",
    payload: babies
  };
}

export function getBabiesFailure(err): Action {
  return {
    type: "GET_BABIES_FAILURE",
    payload: err,
    error: true
  };
}

export function getThisWeeksActivitiesSuccess(babies): Action {
  return {
    type: "GET_THIS_WEEKS_ACTIVITIES_SUCCESS",
    payload: babies
  };
}

export function getThisWeeksActivitiesFailure(err): Action {
  return {
    type: "GET_THIS_WEEKS_ACTIVITIES_FAILURE",
    payload: err,
    error: true
  };
}

const createBabyEpic = (action$: any) => {
  return action$
    .filter((action: MutationResultAction) => {
      return (
        action.type === "APOLLO_MUTATION_RESULT" &&
        action.operationName === "CreateBaby"
      );
    })
    .mergeMap(({ result: { data } }) => {
      if (data.createBaby) {
        const id = data.createBaby.createdBaby.id;
        //
        return [selectBaby(id), resetNavigation("home")];
      }

      return Observable.of(null);
    });
};

export const epics = [createBabyEpic];
