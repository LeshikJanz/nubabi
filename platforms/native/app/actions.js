// @flow
import type { Deps } from 'core/types';
import { Observable } from 'rxjs/Observable';
import { getBabiesRequest, getBabiesSuccess } from 'core/babies/actions';
import { gql } from 'react-apollo';
import firebase from 'react-native-firebase';
import { appOnline, appError } from 'core/app/actions';

// We will get rid of this once we standarize data fetching
const appOnlineEpic = (action$: any, deps: Deps) => {
  return action$.ofType('ON_AUTH').switchMap(() => {
    const { getState, apollo } = deps;
    const state = getState();

    if (state.auth.isAuthenticated) {
      // TODO: remove from here, this should be a client concern
      // so we should handle on SplashScreen and Profile (initial tab)
      const fetchBabiesQuery = gql`
        query getBabies {
          viewer {
            user {
              id
              avatar {
                url
              }
            }
            babies {
              edges {
                node {
                  id
                  name
                }
              }
            }
          }
        }
      `;

      return Observable.merge(
        Observable.of(getBabiesRequest()),
        // $FlowFixMe
        Observable.zip(
          Observable.fromPromise(apollo.query({ query: fetchBabiesQuery })),
        ).mergeMap(([fetchBabiesResult]) => [
          getBabiesSuccess(
            fetchBabiesResult.data.viewer.babies.edges.map(edge => edge.node),
          ),
          appOnline(true),
        ]),
      ).catch(err => Observable.of(appError(err)));
    }

    return Observable.of(appOnline(true));
  });
};

const authAnalyticsEpic = (action$: any) => {
  return action$
    .filter(action => action.type === 'ON_AUTH' && action.payload.user)
    .switchMap(action => {
      firebase.analytics().setUserId(action.payload.user.uid);
      firebase
        .analytics()
        .setUserProperty('user_email', action.payload.user.email);
      return Observable.of();
    });
};

export const epics = [appOnlineEpic, authAnalyticsEpic];
