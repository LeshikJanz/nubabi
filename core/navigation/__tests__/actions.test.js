import { resetNavigation } from '../actions';

describe('navigation actions', () => {
  describe('resetNavigation', () => {
    it('creates a RESET_NAVIGATION action', () => {
      expect(resetNavigation('home')).toMatchSnapshot();
    });
  });
});
