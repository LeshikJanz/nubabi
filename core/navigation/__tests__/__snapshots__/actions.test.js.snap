// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation actions resetNavigation creates a RESET_NAVIGATION action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "routeName": "home",
  },
  "type": "NAVIGATION_RESET",
}
`;
